core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java

```diff
@@ -211,93 +179,181 @@ public ByteSizeValue translogBufferSize() {
         return availableShards;
     }
 
-    /** returns true if shard exists and is availabe for updates */
-    protected boolean shardAvailable(IndexShard shard) {
-        // shadow replica doesn't have an indexing buffer
-        return shard.canIndex() && CAN_UPDATE_INDEX_BUFFER_STATES.contains(shard.state());
+    /** returns how much heap this shard is using for its indexing buffer */
+    protected long getIndexBufferRAMBytesUsed(IndexShard shard) {
+        return shard.getIndexBufferRAMBytesUsed();
     }
 
-    /** set new indexing and translog buffers on this shard.  this may cause the shard to refresh to free up heap. */
-    protected void updateShardBuffers(IndexShard shard, ByteSizeValue shardIndexingBufferSize, ByteSizeValue shardTranslogBufferSize) {
-        try {
-            shard.updateBufferSize(shardIndexingBufferSize, shardTranslogBufferSize);
-        } catch (EngineClosedException | FlushNotAllowedEngineException e) {
-            // ignore
-        } catch (Exception e) {
-            logger.warn("failed to set shard {} index buffer to [{}]", e, shard.shardId(), shardIndexingBufferSize);
-        }
+    /** ask this shard to refresh, in the background, to free up heap */
+    protected void writeIndexingBufferAsync(IndexShard shard) {
+        shard.writeIndexingBufferAsync();
     }
 
-    /** check if any shards active status changed, now. */
+    /** used by tests to check if any shards active status changed, now. */
     public void forceCheck() {
         statusChecker.run();
     }
 
+    /** called by IndexShard to record that this many bytes were written to translog */
+    public void bytesWritten(int bytes) {
+        statusChecker.bytesWritten(bytes);
+    }
+
+    /** Asks this shard to throttle indexing to one thread */
+    protected void activateThrottling(IndexShard shard) {
+        shard.activateThrottling();
+    }
+
+    /** Asks this shard to stop throttling indexing to one thread */
+    protected void deactivateThrottling(IndexShard shard) {
+        shard.deactivateThrottling();
+    }
+
+    static final class ShardAndBytesUsed implements Comparable<ShardAndBytesUsed> {
+        final long bytesUsed;
+        final IndexShard shard;
+
+        public ShardAndBytesUsed(long bytesUsed, IndexShard shard) {
+            this.bytesUsed = bytesUsed;
+            this.shard = shard;
+        }
+
+        @Override
+        public int compareTo(ShardAndBytesUsed other) {
+            // Sort larger shards first:
+            return Long.compare(other.bytesUsed, bytesUsed);
+        }
+    }
+
     class ShardsIndicesStatusChecker implements Runnable {
+
+        long bytesWrittenSinceCheck;
+
+        /** Shard calls this on each indexing/delete op */
+        public synchronized void bytesWritten(int bytes) {
+            bytesWrittenSinceCheck += bytes;
```

Comment:
hmm I am not sure if this is relevant but do we want to run this under a lock all the time? I mean we will block for the entire check which can be longish? We could use an AtomicLong here instead and protect the run method with a `ReentrantLock` and only do it if we can acquire the lock with `tryLock`. then just do the check in while loop and we are good?